### ISSUE TITLE
Unreasonable rejection

### ISSUE THREAD

**Issue by casper-pragma** [agda/agda#5726]
I have written the code that agda refuses to accept for the reason that it cannot destructure a value of nondata – which is fallacious, because I do in fact pattern match on a datatype object. 
The code is rather long and I cannot find a way to shrink it to minimal repro. It's placed [here](https://gist.github.com/insignia-magnum/f4e90b87102ed0c308ca8a8950aa9586).

It is strange that in code that I have linked, agda accepts `strict-eq` – which has almost an identical signature –  but rejects `equiv`. This doesn't seem correct to me. Is it a bug?

**Comment by andreasabel** [comment]
Change the first clause fo `equiv` to 
```agda
equiv {c = 1} dot dot pt pt = dot
```
and the error goes away.

Lesson: do not split on `a` in a clause that splits on `(c₁ : U a)`, because `a` will be refined by unification.

Here is your fixed code without the standard library:
```agda
open import Agda.Builtin.Nat renaming (Nat to ℕ)
open import Agda.Builtin.Bool

_∧_ : (a b : Bool) → Bool
true ∧ a = a
false ∧ a = false

data Σ (A : Set)(B : A -> Set) : Set where
  [_∣_] : (a : A) → B a → Σ A B

data _×_ (A : Set)(B : Set) : Set where
  _,_ : A → B → A × B

data Dot : Set where
  pt : Dot

data Either (L R : Set) : Set where
  inl : L → Either L R
  inr : R → Either L R

data Empty : Set where

maxN : ℕ → ℕ → ℕ
maxN a b = aux a b
  where
    aux : ℕ → ℕ → ℕ
    aux (suc l) (suc r) = aux l r
    aux zero zero = a
    aux zero (suc _) = b
    aux (suc _) zero = a

mutual
  data U : ℕ → Set where
    nil : U 0
    dot : U 1
    eit prod : {a b : ℕ} → U a → U b → U (maxN a (maxN b 2))
    sig : {a b : ℕ} → (i : U a) → (F[ i ] → U b) → U (maxN a (maxN b 2))
    pi : {a b : ℕ} → (i : U a) → (F[ i ] → U b) → U (maxN a (maxN b 3))

  F[_] : {a : ℕ} → U a → Set
  F[ nil ] = Empty
  F[ dot ] = Dot
  F[ eit lt rt ] = Either F[ lt ] F[ rt ]
  F[ prod lt rt ] = F[ lt ] × F[ rt ]
  F[ sig head spine ] = Σ F[ head ] λ i → F[ spine i ]
  F[ pi head spine ] = (a : F[ head ]) → F[ spine a ]


strict-eq : {a b : ℕ} → (t₁ : U a) → (t₂ : U b) → F[ t₁ ] → F[ t₂ ] → Bool
strict-eq dot dot pt pt = true
strict-eq (eit a b) (eit c d) (inl e) (inl f) = strict-eq a c e f
strict-eq (eit a b) (eit c d) (inr e) (inr f) = strict-eq b d e f
strict-eq (prod a b) (prod c d) (e , f) (g , h) = (strict-eq a c e g) ∧ (strict-eq b d f h)
strict-eq (sig h₁ s₁) (sig h₂ s₂) [ v₁ ∣ w₁ ] [ v₂ ∣ w₂ ] =
  (strict-eq h₁ h₂ v₁ v₂) ∧ (strict-eq (s₁ v₁) (s₂ v₂) w₁ w₂)
strict-eq _ _ _ _ = false


{-# NON_COVERING #-}
{-# TERMINATING #-}
equiv : {a b c : ℕ} → (c₁ : U a) → (c₂ : U b) → F[ c₁ ] → F[ c₂ ] → U c
-- equiv {0}{0}{0} nil nil _ _ = nil
equiv {c = 1} dot dot pt pt = dot
equiv {2}{2}{_} t₁ t₂ v₁ v₂ with strict-eq t₁ t₂ v₁ v₂
equiv {2}{2}{1} _ _ _ _ | true = dot
equiv {2}{2}{3} t₁ t₂ v₁ v₂ | false =
  sig (
    sig (pi t₁ λ _ → t₂) (λ f → sig (pi t₂ λ _ → t₁) (λ f⁻¹ →
    prod (equiv t₁ t₁ (f⁻¹ (f v₁)) v₁) (equiv t₂ t₂ (f (f⁻¹ v₂)) v₂))))
  λ { [ _ ∣ [ _ ∣ w ] ] → equiv (prod dot dot)(prod dot dot) w (pt , pt) }
equiv {_}{_}{0} _ _ _ _ = nil
```

**Comment by casper-pragma** [comment]
Thanks. I guess, I'll learn to live with that. There's another issue that has shown up right after this one. 
With `equiv` now written like this ...
```agda
{-# NON_COVERING #-}
{-# TERMINATING #-}
equiv : {a b c : ℕ} → (c₁ : U a) → (c₂ : U b) → F[ c₁ ] → F[ c₂ ] → U c
equiv {2}{2}{_} t₁ t₂ v₁ v₂ with strict-eq t₁ t₂ v₁ v₂
equiv {c = 1} _ _ _ _ | true = dot
equiv {2}{2}{3} t₁ t₂ v₁ v₂ | false =
    sig (pi t₁ λ _ → t₂) (λ f → sig (pi t₂ λ _ → t₁) (λ f⁻¹ →
    prod (equiv t₁ t₁ (f⁻¹ (f v₁)) v₁) (equiv t₂ t₂ (f (f⁻¹ v₂)) v₂)))
equiv {c = 0} _ _ _ _ = nil
```
... agda errors at me that it cant solve constraints. Can those be postponed until `equiv` is applied to arguments?

```
Failed to solve the following constraints:
  fiddling-around.aux _a_130 (maxN _b_131 2) _a_130 (maxN _b_131 2)
    = _b_127
    : ℕ
    (blocked on _b_127)
  fiddling-around.aux (maxN 2 (maxN 2 3)) (maxN _b_127 2) 3
  (fiddling-around.aux _b_127 2 _b_127 2)
    = _b_123
    : ℕ
    (blocked on _b_123)
  fiddling-around.aux (maxN 2 (maxN 2 3)) (maxN _b_123 2) 3
  (fiddling-around.aux _b_123 2 _b_123 2)
    = 3
    : ℕ
    (blocked on _b_123)
```

**Comment by andreasabel** [comment]
I think these constraints would be more readable if you defined `maxN` directly by pattern matching and not via `aux` (I do not see what purpose this indirection should serve).

**Comment by casper-pragma** [comment]
I use `maxN` to pick the biggest number and aux allows me to recurse on data and have access to the original values. Is there a way to do it better?

**Comment by andreasabel** [comment]
Instead of `maxN a b = aux a b` you could simply implement `maxN` by pattern matching directly (using the patterns you use for `aux`).
